### ISSUE TITLE
[Feature request] add selfhosted webinterface for jdownloader

### ISSUE THREAD

**Issue by milahu** [jlesage/docker-jdownloader-2#224]
### Idea

the java gui is too ugly / too heavy for my taste
and the VNC tunnel does not make it better

so i scraped the webapp at [my.jdownloader.org](https://my.jdownloader.org/)
and made it work on localhost in my [myjdownloader](https://github.com/milahu/myjdownloader)
to get a selfhosted webinterface for jdownloader

see also
[My.JDownloader over localhost](https://board.jdownloader.org/showthread.php?t=84428)
[local web interface on LAN](https://board.jdownloader.org/showthread.php?t=69825)

one problem is that my.jdownloader.org is closed-source
so this would require some maintenance to keep it working

im running jdownloader in bubblewrap
which is more lightweight than docker

<details>
<summary>
jdownloader-install-and-run.sh
</summary>

```sh
#!/usr/bin/env bash

dir="$(dirname "$0")"

src="https://github.com/milahu/jdownloader-installed"

# TODO better? ~/.cache? ~/.local? ~/.jdownloader?
dst="$HOME/.config/jdownloader"

if ! [ -e "$dst" ]; then
  echo "creating $dst"
  git clone --depth=1 "$src" "$dst"
else
  echo "keeping $dst"
fi

# patch jdownloader
# dont require MyJDownloader login for headless mode
# so we can run our own MyJDownloader service
"$dir"/jdownloader-patch-core.sh "$dst"

# https://my.jdownloader.org/developers/
# My.JDownloader API Documentation
# Pro Tip: It's possible to access the JDownloader API directly (Bypass our server)
# by enabling the so called 'Deprecated API' in the Advanced Options.
# -> jd will listen on port 3128 -> http://localhost:3128/help
cfg="$dst"/cfg/org.jdownloader.api.RemoteAPIConfig.json
if [ "$(cat "$cfg" | jq -r .deprecatedapienabled)" = "false" ]; then
  cat "$cfg" | jq -c '. * { "deprecatedapienabled": true }' | sponge "$cfg"
fi

jre=$(nix-build --no-out-link '<nixpkgs>' -A jre)

bubblewrap=$(nix-build --no-out-link '<nixpkgs>' -A bubblewrap)

a=($bubblewrap/bin/bwrap)

# FIXME add only required /nix/store paths
a+=(--ro-bind /nix /nix)

# TODO better
a+=(--proc /proc)
a+=(--dev /dev)
a+=(--unshare-pid) # ?
a+=(--new-session) # ?
a+=(--unshare-all)
a+=(--share-net)
a+=(--ro-bind /etc/resolv.conf /etc/resolv.conf)
a+=(--die-with-parent)
#a+=(--dir /run/user/$UID) # tmpfs
#a+=(--dir "$dst") # give access to JDownloader.jar etc
a+=(--bind "$dst" "$dst") # give access to JDownloader.jar etc
a+=(--bind "$HOME/Downloads/jdownloader" "/output") # give access to JDownloader.jar etc
#a+=()

a+=($jre/bin/java)

# see also
# https://aur.archlinux.org/cgit/aur.git/tree/JDownloaderHeadless?h=jdownloader2
# headless mode requires username and password in
# $dst/cfg/org.jdownloader.api.myjdownloader.MyJDownloaderSettings.json
a+=(-Djava.awt.headless=true)

a+=(-jar "$dst/JDownloader.jar")

# no. with "-norestart" jd does not restart after update
#a+=(-norestart)

set -x
exec "${a[@]}"
```

</details>

running jdownloader in headless mode
without a my.jdownloader.org account
requires binary patching of the `Core.jar` file

<details>
<summary>
jdownloader-patch-core.sh
</summary>

```sh
#!/usr/bin/env bash

# patch jdownloader
# dont require MyJDownloader login for headless mode

# based on
# https://github.com/NixOS/nixpkgs/pull/136998/files#r720864456

set -eux

recaf=$(NIXPKGS_ALLOW_UNFREE=1 nix-build --no-out-link '<nixpkgs>' -A nur.repos.milahu.recaf-bin)
export PATH="$recaf/bin:$PATH"

now=$(date --utc +%Y%m%dT%H%M%SZ)

jdownloader_path="$HOME/.config/jdownloader"
if [ $# != 0 ]; then
  jdownloader_path="$1"
fi

jar_path="$jdownloader_path/Core.jar"
bak_jar_path="$jar_path.bak.$now"

class="org/jdownloader/api/myjdownloader/MyJDownloaderController"
method="isAlwaysConnectRequired()Z"

# note: no whitespace in paths
temp_base=/run/user/$UID/recaf-jdownloader-MyJDownloaderController-isAlwaysConnectRequired
asm_path=$temp_base.asm
recaf_script_path=$temp_base.recaf
temp_jar_path=$temp_base.patched.jar

# disassemble the method to patch
cat >$recaf_script_path <<EOF
disassemble --destination=$asm_path $class $method
EOF
recaf --input="$jar_path" --script=$recaf_script_path

# check if method is patched
if ! grep -q INVOKESTATIC $asm_path; then
  echo "patch was already applied"
  # cleanup
  # TODO trap exit
  rm $recaf_script_path
  rm $asm_path
  exit
fi

# patch the assembly code
# -INVOKESTATIC org/appwork/utils/Application.isHeadless()Z
# -IRETURN
# +EXPR return false;
sed -i '/INVOKESTATIC/d; s/IRETURN/EXPR return false;/' $asm_path

# assemble the patched method
cat >$recaf_script_path <<EOF
assemble $class $method $asm_path
export $temp_jar_path
EOF
recaf --input="$jar_path" --script=$recaf_script_path

echo "replacing $jar_path"
mv "$jar_path" "$bak_jar_path"
mv $temp_jar_path "$jar_path"

# cleanup
# TODO trap exit
rm $recaf_script_path
rm $asm_path
```

</details>

**Comment by cRaZy-bisCuiT** [comment]
Yes, please make that happen! We could push some Docker-Compose Setup then to host both in one shot. :)

**Comment by unvency** [comment]
Hat einer von euch beiden seit September 2024 JD erfolgreich headless/im Serverbetrieb? Bzw. falls @milahu es geschafft hat: funktioniert es noch immer? Das AppWork-Team ist seit Jahren vor allem damit beschäftigt, die vielen Hoster-Plugins am Laufen zu halten. Große Funktionsänderungen kommen nicht hinzu und auch myJD ist seit >10 Jahren gleich - da hat sich nichts verändert. Wenn dein Patch noch funktioniert würde ich ihn gerne mal testen. Leider hast du ihn nicht als Repo.

Ich habe versucht ein Plugin zu schreiben, sodass core.jar-Patch unnötig wäre; leider bekomme ich es nicht kompiliert. Soll ich es mal als Repo online stellen?


**Have either of you successfully run JD in headless mode or on a server since September 2024? Or, if @milahu managed it: is it still working? For years, the AppWork team has been primarily occupied with keeping the many hoster plugins running. There haven't been any major functional changes, and myJD has remained the same for over a decade—nothing has changed there. If your patch still works, I’d like to test it out. Unfortunately, you don't have it in a repository.

I tried writing a plugin to make the core.jar patch unnecessary, but unfortunately, I can't get it to compile. Should I put it online as a repository?**

**Comment by milahu** [comment]
> is it still working

i dont know, i dont use jdownloader, i use [pyload](https://github.com/milahu/pyload)

> If your patch still works, I’d like to test it out. Unfortunately, you don't have it in a repository.

maybe `jdownloader-patch-core.sh` is no longer be necessary...

https://support.jdownloader.org/en/tickets/view/UHQC9555-SWRO-2460VASP?token=70d845f88b942f97b9c94af4779ecbbe01725863

<blockquote>

jiaz
Staff - 26/09/2024 2:58 PM

With next update in

cfg/org.jdownloader.api.RemoteAPIConfig.json

deprecatedapienabled true

headlessmyjdownloadermandatory false

</blockquote>

https://support.jdownloader.org/en/tickets/view/NQRJ5131-SLGN-5839MUUI?token=773adc938ebc17004501b56c063fb0df63093d96

<blockquote>

jiaz
Staff - 26/09/2024 3:44 PM

"having no console on stdin should be an expected situation"

-> we do not see this as an expected situation because it is a not supported use case.

Either you have working MyJDownloader or console. Pure headless support is not available.

There are other places that can require console, eg confirmation dialogs and without MyJDownloader

nor a console, you cannot answer those dialogs as well.

</blockquote>

you still need to enable the hidden `Deprecated API` feature

> jdownloader has a hidden `Deprecated API` feature
>
> ```
> dst="$HOME/.config/jdownloader"
> cfg="$dst"/cfg/org.jdownloader.api.RemoteAPIConfig.json
> if [ "$(cat "$cfg" | jq -r .deprecatedapienabled)" = "false" ]; then
>   cat "$cfg" | jq -c '. * { "deprecatedapienabled": true }' | sponge "$cfg"
> fi
> ```

you can scrape the latest version of the webui with my [webapps-scraper](https://github.com/milahu/webapps-scraper)

> the initial version was created with [webapps-scraper](https://github.com/milahu/webapps-scraper) from [my.jdownloader.org](https://my.jdownloader.org/)

[let me know](https://github.com/milahu/contact) if you need more help

**Comment by hexxone** [comment]
> **Have either of you successfully run JD in headless mode or on a server since September 2024? Or, if [@milahu](https://github.com/milahu) managed it: is it still working? For years, the AppWork team has been primarily occupied with keeping the many hoster plugins running. There haven't been any major functional changes, and myJD has remained the same for over a decade—nothing has changed there. If your patch still works, I’d like to test it out. Unfortunately, you don't have it in a repository.
> 
> I tried writing a plugin to make the core.jar patch unnecessary, but unfortunately, I can't get it to compile. Should I put it online as a repository?**

No, unfortunately as milahu mentioned, I also dont have a way to run jdownloader2 with true "headless" mode..

I am now using myJd Account instead and have settled for automation via API "remotely" rather than relying on the clunky JD2 Web interface and the myJD Browser Extension (which also ceased to work some days ago completely due to Manifest V3)...
https://github.com/hexxone/jdownloader2 is the image I'm running successfully to this day.

I also tried pyload, but unfortunately its still doesnt support the automatic captcha solvers and a lot of download sites/plugins which I often use..

**Comment by hexxone** [comment]
Update @unvency  ! I could confirm what @milahu said. 

the 3 settings together are now actually working for true headless mode - I just tested with my own repo above:
```json
{
  "deprecatedapienabled": true,
  "headlessmyjdownloadermandatory": false,
  "deprecatedapilocalhostonly": false
}
```
with those 3, you can actually successfully use the latest jdownloader2 deprecated api, headless without a myJD-Account in Docker and externally connect to it from your host.

**Comment by milahu** [comment]
yay : )
but does [my webui](https://github.com/milahu/myjdownloader) still work with the latest jdownloader?
